datamodel/test/core_tests.cpp

```diff
@@ -0,0 +1,63 @@
+#include <gtest/gtest.h>
+
+#include <rcbe-engine/datamodel/core/CoreObject.hpp>
+#include <rcbe-engine/datamodel/core/Dimensions.hpp>
+#include <rcbe-engine/fundamentals/types.hpp>
+
+inline constexpr rcbe::core::EngineIntergral IWIDTH = 1024;
+inline constexpr rcbe::core::EngineIntergral IHEIGHT = 768;
+
+namespace {
+struct some_empty_struct {};
+}
+
+TEST(DimensionsTests, IntegerDimensions) {
+    rcbe::core::Dimensions dim{IWIDTH, IHEIGHT};
+
+    ASSERT_EQ(dim.width, IWIDTH);
+    ASSERT_EQ(dim.height, IHEIGHT);
+}
+
+TEST(DimensionsTests, FloatDimensions) {
+    using FloatDimensions = rcbe::core::dimensions_impl<float>;
+    static constexpr float FWIDTH = static_cast<float>(IWIDTH);
+    static constexpr float FHEIGHT = static_cast<float>(IHEIGHT);
+    FloatDimensions fdim{FWIDTH, FHEIGHT};
+
+    ASSERT_EQ(fdim.width, FWIDTH);
+    ASSERT_EQ(fdim.height, FHEIGHT);
+}
+
+TEST(DimensionsTests, DoubleDimensions) {
+    static constexpr float DWIDTH = static_cast<rcbe::core::EngineScalar>(IWIDTH);
+    static constexpr float DHEIGHT = static_cast<rcbe::core::EngineScalar>(IHEIGHT);
+    rcbe::core::ScalarDimensions sdim{DWIDTH, DHEIGHT};
+
+    ASSERT_EQ(sdim.width, DWIDTH);
+    ASSERT_EQ(sdim.height, DHEIGHT);
+}
+
+TEST(CoreObjectTests, Components) {
+    using DimT = rcbe::core::Dimensions;
+
+    rcbe::core::CoreObject co{some_empty_struct{}};
+
+    DimT dim{IWIDTH, IHEIGHT};
+    co.addComponent(std::move(dim));
+
+    ASSERT_TRUE(co.hasComponent<DimT>());
+    const auto dim_comp_ptr = co.getComponent<DimT>();
+    ASSERT_NE(dim_comp_ptr, nullptr);
+    
+    const auto &dim_comp = dim_comp_ptr->as<DimT>();
```

Comment:
Now, I think making user get component via it's type, and then explicitly cast to that type using one more operation is a code smell. Consider at least leaving a `TODO` to think how it should be fixed in the future.

Reply:
Say user wants to get component to view it's tags. In this case they don't need casting. So returning pointer as is, and leaving the `as` invocation to user seems fine.

Reply:
Ok. sounds reasonable.